workflow/controller/operator.go

```diff
@@ -2683,15 +2683,23 @@ func (woc *wfOperationCtx) loadExecutionSpec() (wfv1.TemplateReferenceHolder, wf
 
 	executionParameters := woc.wf.Spec.Arguments
 
-	woc.wfSpec = &woc.wf.Spec
 	if woc.wf.Spec.WorkflowTemplateRef == nil {
+		woc.wfSpec = &woc.wf.Spec
+
+		if woc.controller.referenceMode {
+			return nil, wfv1.Arguments{}, fmt.Errorf("workflows may only use workflowTemplateRef to be executed when the controller is in reference mode")
+		}
+
 		tmplRef := &wfv1.WorkflowStep{Template: woc.wfSpec.Entrypoint}
 		return tmplRef, executionParameters, nil
 	}
 
-	if woc.wf.Status.StoredWorkflowSpec == nil {
+	// If the controller is in reference mode, don't trust the stored spec in the Workflow object and refetch it every time
+	if woc.wf.Status.StoredWorkflowSpec == nil || woc.controller.referenceMode {
```

Comment:
This is susceptible to `WorkflowTemplates` changing between different operations; however allowing the spec to be stored in status would make this change moot as an attacker could simply add their desired code in the status field.